Wild, this triggers the compiler crash.

```
> Task :zipline:linkDebugTestIosSimulatorArm64
e: Compilation failed: Assertion failed

 * Source files: 
 * Compiler version info: Konan: 1.5.30 / Kotlin: 1.5.30
 * Output kind: PROGRAM

e: java.lang.AssertionError: Assertion failed
	at org.jetbrains.kotlin.backend.konan.llvm.DebugUtilsKt.location(DebugUtils.kt:99)
	at org.jetbrains.kotlin.backend.konan.llvm.DebugUtilsKt.line(DebugUtils.kt:108)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$FunctionScope.location(IrToBitcode.kt:688)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$InnerScope.location(IrToBitcode.kt)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$InnerScope.location(IrToBitcode.kt)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$InnerScope.location(IrToBitcode.kt)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$ReturnableBlockScope.location(IrToBitcode.kt:1771)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor$InnerScope.location(IrToBitcode.kt)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.getStartLocation(IrToBitcode.kt:1915)
	at org.jetbrains.kotlin.backend.konan.llvm.CodeGeneratorVisitor.updateBuilderDebugLocation(IrToBitcode.kt:1909)
```

--- a/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/InboundBridgeRewriter.kt
+++ b/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/InboundBridgeRewriter.kt
@@ -20,8 +20,6 @@ import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
 import org.jetbrains.kotlin.backend.common.ir.addFakeOverrides
 import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
 import org.jetbrains.kotlin.backend.common.ir.isSuspend
-import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
-import org.jetbrains.kotlin.descriptors.Modality
 import org.jetbrains.kotlin.ir.builders.IrBlockBodyBuilder
 import org.jetbrains.kotlin.ir.builders.IrBuilderWithScope
 import org.jetbrains.kotlin.ir.builders.declarations.addConstructor
@@ -40,7 +38,6 @@ import org.jetbrains.kotlin.ir.builders.irReturn
 import org.jetbrains.kotlin.ir.builders.irString
 import org.jetbrains.kotlin.ir.builders.irWhen
 import org.jetbrains.kotlin.ir.declarations.IrClass
-import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
 import org.jetbrains.kotlin.ir.declarations.IrDeclarationParent
 import org.jetbrains.kotlin.ir.declarations.IrFactory
 import org.jetbrains.kotlin.ir.declarations.IrProperty
@@ -61,7 +58,6 @@ import org.jetbrains.kotlin.ir.symbols.IrFunctionSymbol
 import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
 import org.jetbrains.kotlin.ir.types.defaultType
 import org.jetbrains.kotlin.ir.types.typeWith
-import org.jetbrains.kotlin.ir.util.SYNTHETIC_OFFSET
 import org.jetbrains.kotlin.ir.util.constructors
 import org.jetbrains.kotlin.ir.util.defaultType
 import org.jetbrains.kotlin.ir.util.getPropertyGetter
@@ -194,8 +190,7 @@ internal class InboundBridgeRewriter(
   private fun irNewInboundBridge(): IrContainerExpression {
     val inboundBridgeOfT = ziplineApis.inboundBridge.typeWith(bridgedInterface.type)
     val inboundBridgeSubclass = irFactory.buildClass {
-      name = Name.special("<no name provided>")
-      visibility = DescriptorVisibilities.LOCAL
+      initDefaults(original)
     }.apply {
       superTypes = listOf(inboundBridgeOfT)
       createImplicitParameterDeclarationWithWrappedDescriptor()
@@ -204,9 +199,7 @@ internal class InboundBridgeRewriter(
     // InboundBridge<EchoService>()
     val superConstructor = ziplineApis.inboundBridge.constructors.single()
     val constructor = inboundBridgeSubclass.addConstructor {
-      origin = IrDeclarationOrigin.DEFINED
-      visibility = DescriptorVisibilities.PUBLIC
-      isPrimary = true
+      initDefaults(original)
     }.apply {
       irConstructorBody(pluginContext) { statements ->
         statements += irDelegatingConstructorCall(
@@ -236,8 +229,8 @@ internal class InboundBridgeRewriter(
     )
 
     return IrBlockBodyBuilder(
-      startOffset = SYNTHETIC_OFFSET,
-      endOffset = SYNTHETIC_OFFSET,
+      startOffset = original.startOffset,
+      endOffset = original.endOffset,
       context = pluginContext,
       scope = scope.scope,
     ).irBlock(origin = IrStatementOrigin.OBJECT_LITERAL) {
@@ -255,15 +248,16 @@ internal class InboundBridgeRewriter(
     // override fun create(context: Context): InboundCallHandler {
     // }
     val result = inboundBridgeSubclass.addFunction {
+      initDefaults(original)
       name = ziplineApis.inboundBridgeCreate.owner.name
-      visibility = DescriptorVisibilities.PUBLIC
-      modality = Modality.OPEN
       returnType = ziplineApis.inboundCallHandler.defaultType
     }.apply {
       addDispatchReceiver {
+        initDefaults(original)
         type = inboundBridgeSubclass.defaultType
       }
       addValueParameter {
+        initDefaults(original)
         name = Name.identifier("context")
         type = ziplineApis.inboundBridgeContext.defaultType
       }
@@ -293,17 +287,14 @@ internal class InboundBridgeRewriter(
     // }
     val inboundCallHandler = ziplineApis.inboundCallHandler.defaultType
     val inboundCallHandlerSubclass = irFactory.buildClass {
-      name = Name.special("<no name provided>")
-      visibility = DescriptorVisibilities.LOCAL
+      initDefaults(original)
     }.apply {
       superTypes = listOf(inboundCallHandler)
       createImplicitParameterDeclarationWithWrappedDescriptor()
     }
 
     val constructor = inboundCallHandlerSubclass.addConstructor {
-      origin = IrDeclarationOrigin.DEFINED
-      visibility = DescriptorVisibilities.PUBLIC
-      isPrimary = true
+      initDefaults(original)
     }.apply {
       irConstructorBody(pluginContext) { statements ->
         statements += irDelegatingConstructorCall(
@@ -357,8 +348,8 @@ internal class InboundBridgeRewriter(
     }
 
     return IrBlockBodyBuilder(
-      startOffset = SYNTHETIC_OFFSET,
-      endOffset = SYNTHETIC_OFFSET,
+      startOffset = original.startOffset,
+      endOffset = original.endOffset,
       context = pluginContext,
       scope = scope.scope,
     ).irBlock(origin = IrStatementOrigin.OBJECT_LITERAL) {
@@ -380,16 +371,17 @@ internal class InboundBridgeRewriter(
       else -> ziplineApis.inboundCallHandlerCall
     }
     return inboundCallHandler.addFunction {
+      initDefaults(original)
       name = inboundBridgeCall.owner.name
-      visibility = DescriptorVisibilities.PUBLIC
-      modality = Modality.OPEN
       returnType = ziplineApis.stringArrayType
       isSuspend = callSuspending
     }.apply {
       addDispatchReceiver {
+        initDefaults(original)
         type = inboundCallHandler.defaultType
       }
       addValueParameter {
+        initDefaults(original)
         name = Name.identifier("inboundCall")
         type = ziplineApis.inboundCall.defaultType
       }

--- a/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/OutboundBridgeRewriter.kt
+++ b/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/OutboundBridgeRewriter.kt
@@ -19,9 +19,6 @@ import org.jetbrains.kotlin.backend.common.ScopeWithIr
 import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
 import org.jetbrains.kotlin.backend.common.ir.addFakeOverrides
 import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
-import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
-import org.jetbrains.kotlin.descriptors.Modality
-import org.jetbrains.kotlin.ir.util.SYNTHETIC_OFFSET
 import org.jetbrains.kotlin.ir.builders.IrBlockBodyBuilder
 import org.jetbrains.kotlin.ir.builders.declarations.addConstructor
 import org.jetbrains.kotlin.ir.builders.declarations.addDispatchReceiver
@@ -145,8 +142,7 @@ internal class OutboundBridgeRewriter(
   private fun irNewOutboundBridge(): IrContainerExpression {
     val outboundBridgeOfT = ziplineApis.outboundBridge.typeWith(bridgedInterface.type)
     val outboundBridgeSubclass = irFactory.buildClass {
-      name = Name.special("<no name provided>")
-      visibility = DescriptorVisibilities.LOCAL
+      initDefaults(original)
     }.apply {
       superTypes = listOf(outboundBridgeOfT)
       createImplicitParameterDeclarationWithWrappedDescriptor()
@@ -155,9 +151,7 @@ internal class OutboundBridgeRewriter(
     // OutboundBridge<EchoService>()
     val superConstructor = ziplineApis.outboundBridge.constructors.single()
     val constructor = outboundBridgeSubclass.addConstructor {
-      origin = IrDeclarationOrigin.DEFINED
-      visibility = DescriptorVisibilities.PUBLIC
-      isPrimary = true
+      initDefaults(original)
     }.apply {
       irConstructorBody(pluginContext) { statements ->
         statements += irDelegatingConstructorCall(
@@ -178,15 +172,16 @@ internal class OutboundBridgeRewriter(
     // override fun create(callFactory: OutboundCall.Factory): EchoService {
     // }
     val createFunction = outboundBridgeSubclass.addFunction {
+      initDefaults(original)
       name = Name.identifier("create")
-      visibility = DescriptorVisibilities.PUBLIC
-      modality = Modality.OPEN
       returnType = bridgedInterface.type
     }.apply {
       addDispatchReceiver {
+        initDefaults(original)
         type = outboundBridgeSubclass.defaultType
       }
       addValueParameter {
+        initDefaults(original)
         name = Name.identifier("context")
         type = ziplineApis.outboundBridgeContext.defaultType
       }
@@ -204,8 +199,8 @@ internal class OutboundBridgeRewriter(
     }
 
     return IrBlockBodyBuilder(
-      startOffset = SYNTHETIC_OFFSET,
-      endOffset = SYNTHETIC_OFFSET,
+      startOffset = original.startOffset,
+      endOffset = original.endOffset,
       context = pluginContext,
       scope = scope.scope,
     ).irBlock(origin = IrStatementOrigin.OBJECT_LITERAL) {
@@ -223,8 +218,7 @@ internal class OutboundBridgeRewriter(
     //   ...
     // }
     val clientImplementation = irFactory.buildClass {
-      name = Name.special("<no name provided>")
-      visibility = DescriptorVisibilities.LOCAL
+      initDefaults(original)
     }.apply {
       parent = createFunction
       superTypes = listOf(bridgedInterface.type)
@@ -232,9 +226,7 @@ internal class OutboundBridgeRewriter(
     }
 
     val constructor = clientImplementation.addConstructor {
-      origin = IrDeclarationOrigin.DEFINED
-      visibility = DescriptorVisibilities.PUBLIC
-      isPrimary = true
+      initDefaults(original)
     }
     constructor.irConstructorBody(pluginContext) { statements ->
       statements += irDelegatingConstructorCall(
@@ -276,20 +268,21 @@ internal class OutboundBridgeRewriter(
   ): IrSimpleFunction {
     val functionReturnType = bridgedInterface.resolveTypeParameters(bridgedFunction.returnType)
     val result = addFunction {
+      initDefaults(original)
       name = bridgedFunction.name
-      visibility = DescriptorVisibilities.PUBLIC
-      modality = Modality.OPEN
       isSuspend = bridgedFunction.isSuspend
       returnType = functionReturnType
     }.apply {
       overriddenSymbols = listOf(bridgedFunction.symbol)
       addDispatchReceiver {
+        initDefaults(original)
         type = defaultType
       }
     }
 
     for (valueParameter in bridgedFunction.valueParameters) {
       result.addValueParameter {
+        initDefaults(original)
         name = valueParameter.name
         type = bridgedInterface.resolveTypeParameters(valueParameter.type)
       }

--- a/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/ir.kt
+++ b/zipline-kotlin-plugin-hosted/src/main/kotlin/app/cash/zipline/kotlin/ir.kt
@@ -25,12 +25,14 @@ import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
 import org.jetbrains.kotlin.descriptors.Modality
 import org.jetbrains.kotlin.ir.IrElement
 import org.jetbrains.kotlin.ir.IrStatement
-import org.jetbrains.kotlin.ir.util.SYNTHETIC_OFFSET
 import org.jetbrains.kotlin.ir.builders.IrBlockBodyBuilder
 import org.jetbrains.kotlin.ir.builders.IrBlockBuilder
 import org.jetbrains.kotlin.ir.builders.IrBuilderWithScope
 import org.jetbrains.kotlin.ir.builders.IrGeneratorContext
 import org.jetbrains.kotlin.ir.builders.Scope
+import org.jetbrains.kotlin.ir.builders.declarations.IrClassBuilder
+import org.jetbrains.kotlin.ir.builders.declarations.IrFunctionBuilder
+import org.jetbrains.kotlin.ir.builders.declarations.IrValueParameterBuilder
 import org.jetbrains.kotlin.ir.builders.irGet
 import org.jetbrains.kotlin.ir.builders.irGetField
 import org.jetbrains.kotlin.ir.builders.irReturn
@@ -57,6 +59,7 @@ import org.jetbrains.kotlin.ir.symbols.impl.IrFieldSymbolImpl
 import org.jetbrains.kotlin.ir.symbols.impl.IrPropertySymbolImpl
 import org.jetbrains.kotlin.ir.symbols.impl.IrSimpleFunctionSymbolImpl
 import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.util.SYNTHETIC_OFFSET
 import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.name.Name
 
@@ -100,6 +103,30 @@ internal fun IrBuilderWithScope.irReturn(
   )
 }
 
+/** Set up reasonable defaults for a generated function or constructor. */
+fun IrFunctionBuilder.initDefaults(original: IrElement) {
+  this.startOffset = original.startOffset
+  this.endOffset = original.endOffset
+  this.origin = IrDeclarationOrigin.DEFINED
+  this.visibility = DescriptorVisibilities.PUBLIC
+  this.modality = Modality.OPEN
+  this.isPrimary = true
+}
+
+/** Set up reasonable defaults for a generated class. */
+fun IrClassBuilder.initDefaults(original: IrElement) {
+  this.startOffset = original.startOffset
+  this.endOffset = original.endOffset
+  this.name = Name.special("<no name provided>")
+  this.visibility = DescriptorVisibilities.LOCAL
+}
+
+/** Set up reasonable defaults for a value parameter. */
+fun IrValueParameterBuilder.initDefaults(original: IrElement) {
+  this.startOffset = original.startOffset
+  this.endOffset = original.endOffset
+}
+
 fun IrConstructor.irConstructorBody(
   context: IrGeneratorContext,
   blockBody: DeclarationIrBuilder.(MutableList<IrStatement>) -> Unit
@@ -107,8 +134,8 @@ fun IrConstructor.irConstructorBody(
   val constructorIrBuilder = DeclarationIrBuilder(
     generatorContext = context,
     symbol = IrSimpleFunctionSymbolImpl(),
-    startOffset = SYNTHETIC_OFFSET,
-    endOffset = SYNTHETIC_OFFSET
+    startOffset = startOffset,
+    endOffset = endOffset
   )
   body = context.irFactory.createBlockBody(
     startOffset = constructorIrBuilder.startOffset,
@@ -155,8 +182,8 @@ fun IrSimpleFunction.irFunctionBody(
   blockBody: IrBlockBodyBuilder.() -> Unit
 ) {
   val bodyBuilder = IrBlockBodyBuilder(
-    startOffset = SYNTHETIC_OFFSET,
-    endOffset = SYNTHETIC_OFFSET,
+    startOffset = startOffset,
+    endOffset = endOffset,
     context = context,
     scope = Scope(scopeOwnerSymbol),
   )
@@ -211,8 +238,8 @@ fun irVal(
     parent = declaringClass
     correspondingPropertySymbol = result.symbol
     val initializerBuilder = IrBlockBuilder(
-      startOffset = SYNTHETIC_OFFSET,
-      endOffset = SYNTHETIC_OFFSET,
+      startOffset = declaringClass.startOffset,
+      endOffset = declaringClass.endOffset,
       context = pluginContext,
       scope = Scope(symbol),
     )

